File: LCP-07/07.py
class Solution:
  def numWays(self, n: int, relation: List[List[int]], k: int) -> int:
    dp = [0 for _ in range(n + 1)]
    dp[0] = 1
    for i in range(k):
      next = [0 for _ in range(n + 1)]
      for edge in relation:
        src = edge[0]
        dst = edge[1:]
        next[dst] += dp[src]
      dp = next
    return dp[n - 1]

File: LCP-07/07.java
class Solution {
  public int numWays(int n, int[][] relation, int k) {
    int[] dp = new int[n];
    dp[0] = 1;
    for (int i = 0; i < k; i++) {
      int[] next = new int[n];
      for (int[] edge : relation) {
        int src = edge[0], dst = edge[1];
        next[dst] += dp[src];
      }
      dp = next;
    }
    return dp[n - 1];
  }
}